packages/utils/src/syncpromise.ts

```diff
@@ -0,0 +1,408 @@
+import { isThenable } from './is';
+
+/** JSDoc */
+enum States {
+  /** JSDoc */
+  PENDING = 'PENDING',
+  /** JSDoc */
+  RESOLVED = 'RESOLVED',
+  /** JSDoc */
+  REJECTED = 'REJECTED',
+}
+
+/** JSDoc */
+interface Handler<T, U> {
+  onSuccess: HandlerOnSuccess<T, U>;
+  onFail: HandlerOnFail<U>;
+}
+
+type HandlerOnSuccess<T, U = any> = (value: T) => U | Thenable<U>;
+type HandlerOnFail<U = any> = (reason: any) => U | Thenable<U>;
+type Finally<U> = () => U | Thenable<U>;
+
+/** JSDoc */
+interface Thenable<T> {
+  then<U>(onSuccess?: HandlerOnSuccess<T, U>, onFail?: HandlerOnFail<U> | ((reason: any) => void)): Thenable<U>;
+}
+
+type Resolve<R> = (value?: R | Thenable<R> | any) => void;
+type Reject = (value?: any) => void;
+
+/** JSDoc */
+export class SyncPromise<T> {
+  /** JSDoc */
+  private state: States = States.PENDING;
+  /** JSDoc */
+  private handlers: Array<Handler<T, any>> = [];
+  /** JSDoc */
+  private value: T | any;
+
+  public constructor(callback: (resolve: Resolve<T>, reject: Reject) => void) {
+    try {
+      callback(this.resolve, this.reject);
+    } catch (e) {
+      this.reject(e);
+    }
+  }
+
+  /** JSDoc */
+  private readonly resolve = (value: T) => {
+    this.setResult(value, States.RESOLVED);
+  };
+
+  /** JSDoc */
+  private readonly reject = (reason: any) => {
+    this.setResult(reason, States.REJECTED);
+  };
+
+  /** JSDoc */
+  private readonly setResult = (value: T | any, state: States) => {
+    if (this.state !== States.PENDING) {
+      return null;
+    }
+
+    if (isThenable(value)) {
+      return (value as Thenable<T>).then(this.resolve, this.reject);
+    }
+
+    this.value = value;
+    this.state = state;
+
+    return this.executeHandlers();
+  };
+
+  /** JSDoc */
+  private readonly executeHandlers = () => {
+    if (this.state === States.PENDING) {
+      return null;
+    }
+
+    this.handlers.forEach(handler => {
+      if (this.state === States.REJECTED) {
+        // tslint:disable-next-line:no-unsafe-any
+        return handler.onFail(this.value);
+      }
+      // tslint:disable-next-line:no-unsafe-any
+      return handler.onSuccess(this.value);
+    });
+
+    this.handlers = [];
+    return null;
+  };
+
+  /** JSDoc */
+  private readonly attachHandler = (handler: Handler<T, any>) => {
+    this.handlers = this.handlers.concat(handler);
```

Comment:
This one is odd, as attaching handler always executes it anyway. What's the point of keeping track of them then?